File: frontend/src/lib/api/analytics.ts
import { request } from "@/lib/api-client"

export type AnalyticsComparison = {
  value: number
  previous_value: number
  change_percent: number | null
}

export type AnalyticsNullableComparison = {
  value: number | null
  previous_value: number | null
  change_percent: number | null
}

export type WorkspaceAnalytics = {
  summary: {
    members: { total: number; active: number }
    active_teams: number
    active_users: AnalyticsComparison
    runs: AnalyticsComparison
    tokens: {
      input: number
      output: number
      application_total: number
      graph_total: number
      total: number
      unreported_runs: number
      unreported_graph_builds: number
      previous_total: number
      change_percent: number | null
    }
    success_rate: AnalyticsNullableComparison
    average_duration_ms: AnalyticsNullableComparison
  }
  trends: Array<{
    date: string
    runs: number
    graph_builds: number
    input_tokens: number
    output_tokens: number
    application_tokens: number
    graph_tokens: number
    total_tokens: number
  }>
  hourly_runs: Array<{
    hour: number
    runs: number
  }>
  distributions: {
    run_types: Array<{ key: string; count: number }>
    access_sources: Array<{ key: string; count: number }>
    statuses: Array<{ key: string; count: number }>
  }
  rankings: {
    users: Array<{
      user_id: string
      name: string
      run_count: number
      total_tokens: number
    }>
    applications: Array<{
      application_id: string
      name: string
      app_type: "agent" | "workflow"
      run_count: number
      total_tokens: number
      success_rate: number | null
    }>
    anonymous: { run_count: number; total_tokens: number }
    teams: Array<{
      team_id: string
      name: string
      peak_daily_runs: number
      run_count: number
    }>
  }
  frequent_questions: Array<{
    question: string
    count: number
    latest_at: string
  }>
  metadata: {
    workspace_id: string
    timezone: "UTC"
    from_date: string
    to_date: string
    previous_from_date: string
    previous_to_date: string
    end_exclusive: true
    generated_at: string
  }
}

export function getWorkspaceAnalytics(
  token: string,
  workspaceId: string,
  range: { from: string; to: string },
  signal?: AbortSignal
) {
  const query = new URLSearchParams({ from: range.from, to: range.to })
  return request<WorkspaceAnalytics>(
    `/api/v1/workspaces/${workspaceId}/analytics?${query}`,
    { token, signal }
  )
}

File: backend/app/schemas/analytics.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field


class AnalyticsCountComparison(BaseModel):
    value: int
    previous_value: int
    change_percent: float | None


class AnalyticsRatioComparison(BaseModel):
    value: float | None
    previous_value: float | None
    change_percent: float | None


class WorkspaceAnalyticsMemberSummary(BaseModel):
    total: int
    active: int


class WorkspaceAnalyticsTokenSummary(BaseModel):
    input: int
    output: int
    application_total: int
    graph_total: int
    total: int
    unreported_runs: int
    unreported_graph_builds: int
    previous_total: int
    change_percent: float | None


class WorkspaceAnalyticsSummary(BaseModel):
    members: WorkspaceAnalyticsMemberSummary
    active_teams: int
    active_users: AnalyticsCountComparison
    runs: AnalyticsCountComparison
    tokens: WorkspaceAnalyticsTokenSummary
    success_rate: AnalyticsRatioComparison
    average_duration_ms: AnalyticsRatioComparison


class WorkspaceAnalyticsTrendPoint(BaseModel):
    date: date
    runs: int
    graph_builds: int
    input_tokens: int
    output_tokens: int
    application_tokens: int
    graph_tokens: int
    total_tokens: int


class WorkspaceAnalyticsHourlyPoint(BaseModel):
    hour: int = Field(ge=0, le=23)
    runs: int


class WorkspaceAnalyticsDistributionItem(BaseModel):
    key: str
    count: int


class WorkspaceAnalyticsDistributions(BaseModel):
    run_types: list[WorkspaceAnalyticsDistributionItem]
    access_sources: list[WorkspaceAnalyticsDistributionItem]
    statuses: list[WorkspaceAnalyticsDistributionItem]


class WorkspaceAnalyticsUserRankingItem(BaseModel):
    user_id: str
    name: str
    run_count: int
    total_tokens: int


class WorkspaceAnalyticsApplicationRankingItem(BaseModel):
    application_id: str
    name: str
    app_type: Literal["agent", "workflow"]
    run_count: int
    total_tokens: int
    success_rate: float | None


class WorkspaceAnalyticsAnonymousUsage(BaseModel):
    run_count: int
    total_tokens: int


class WorkspaceAnalyticsTeamRankingItem(BaseModel):
    team_id: str
    name: str
    peak_daily_runs: int
    run_count: int


class WorkspaceAnalyticsRankings(BaseModel):
    users: list[WorkspaceAnalyticsUserRankingItem]
    applications: list[WorkspaceAnalyticsApplicationRankingItem]
    anonymous: WorkspaceAnalyticsAnonymousUsage
    teams: list[WorkspaceAnalyticsTeamRankingItem]


class WorkspaceAnalyticsFrequentQuestion(BaseModel):
    question: str
    count: int
    latest_at: datetime


class WorkspaceAnalyticsMetadata(BaseModel):
    workspace_id: str
    timezone: Literal["Asia/Shanghai"]
    from_date: date
    to_date: date
    previous_from_date: date
    previous_to_date: date
    end_exclusive: Literal[True]
    generated_at: datetime


class WorkspaceAnalyticsResponse(BaseModel):
    summary: WorkspaceAnalyticsSummary
    trends: list[WorkspaceAnalyticsTrendPoint]
    hourly_runs: list[WorkspaceAnalyticsHourlyPoint]
    distributions: WorkspaceAnalyticsDistributions
    rankings: WorkspaceAnalyticsRankings
    frequent_questions: list[WorkspaceAnalyticsFrequentQuestion]
    metadata: WorkspaceAnalyticsMetadata
